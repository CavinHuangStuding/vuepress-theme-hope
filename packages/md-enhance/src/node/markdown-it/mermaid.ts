/* eslint-disable max-statements */
import hash = require("hash-sum");
import MarkdownIt = require("markdown-it");
import Token = require("markdown-it/lib/token");

const mermaid = (md: MarkdownIt): void => {
  const mermaidRender = (tokens: Token[], idx: number): string => {
    const token = tokens[idx];
    const key = `mermaid${hash(idx)}`;
    const { content } = token;

    return `<Mermaid id="${key}" data-code="${encodeURIComponent(
      content
    )}"></Mermaid>`;
  };

  // Handle ```mermaid blocks
  const fence = md.renderer.rules.fence;

  md.renderer.rules.fence = (...args): string => {
    const [tokens, idx] = args;
    const { info } = tokens[idx];

    if (info.trim().split(":")[0] === "mermaid")
      return mermaidRender(tokens, idx);

    // eslint-disable-next-line @typescript-eslint/no-non-null-assertion
    return fence!(...args);
  };

  md.renderer.rules.mermaid = mermaidRender;
};

export default mermaid;
