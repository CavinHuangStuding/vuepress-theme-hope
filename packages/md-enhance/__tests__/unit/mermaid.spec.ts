import MarkdownIt = require("markdown-it");
import mermaid from "../../src/node/markdown-it/mermaid";

const demo = `flowchart TB
    c1-->a2
    subgraph one
    a1-->a2
    end
    subgraph two
    b1-->b2
    end
    subgraph three
    c1-->c2
    end
    one --> two
    three --> two
    two --> c2`;

describe("mermaid", () => {
  const markdownIt = MarkdownIt({ linkify: true }).use(mermaid);

  it("Should render ```mermaid", () => {
    const renderResult = markdownIt.render(`
\`\`\`mermaid
${demo}
\`\`\`
`);

    expect(renderResult).toMatch(
      /<Mermaid id="mermaid.*?" data-code=".*?"><\/Mermaid>/
    );
    expect(renderResult).toMatchSnapshot();
  });

  it("Shoud not render", () => {
    expect(
      markdownIt.render(`
${demo}
`)
    ).toMatchSnapshot();

    expect(
      markdownIt.render(`
\`\`\`md
${demo}
\`\`\`
`)
    ).toMatchSnapshot();
  });
});
