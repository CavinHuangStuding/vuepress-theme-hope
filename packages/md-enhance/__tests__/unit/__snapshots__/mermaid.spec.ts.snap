// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`mermaid Shoud not render 1`] = `
"<p>flowchart TB
c1--&gt;a2
subgraph one
a1--&gt;a2
end
subgraph two
b1--&gt;b2
end
subgraph three
c1--&gt;c2
end
one --&gt; two
three --&gt; two
two --&gt; c2</p>
"
`;

exports[`mermaid Shoud not render 2`] = `
"<pre><code class=\\"language-md\\">flowchart TB
    c1--&gt;a2
    subgraph one
    a1--&gt;a2
    end
    subgraph two
    b1--&gt;b2
    end
    subgraph three
    c1--&gt;c2
    end
    one --&gt; two
    three --&gt; two
    two --&gt; c2
</code></pre>
"
`;

exports[`mermaid Should render \`\`\`mermaid 1`] = `"<Mermaid id=\\"mermaid1a96284a\\" data-code=\\"flowchart%20TB%0A%20%20%20%20c1--%3Ea2%0A%20%20%20%20subgraph%20one%0A%20%20%20%20a1--%3Ea2%0A%20%20%20%20end%0A%20%20%20%20subgraph%20two%0A%20%20%20%20b1--%3Eb2%0A%20%20%20%20end%0A%20%20%20%20subgraph%20three%0A%20%20%20%20c1--%3Ec2%0A%20%20%20%20end%0A%20%20%20%20one%20--%3E%20two%0A%20%20%20%20three%20--%3E%20two%0A%20%20%20%20two%20--%3E%20c2%0A\\"></Mermaid>"`;
